README.md
# AALeC - AAlener Lern-Computer

## Beschreibung

AALeC ist der Aalener Lerncomputer. Er basiert auf einem ESP8266, welcher als
Wemos-D1-Mini-Modul eingesetzt wird.

Sensoren:
* DHT11-Umweltsensor für Temperatur und Luftfeuchte
* Dreh- /  Drückgeber
* Analoges Potentiometer

Aktoren / Ausgabeelemente:
* Drei Fullcolor-Leuchtdioden PL9823 (WS2812B-kompatibel)
* OLED-Display
* Lautsprecher
* LED einfach

Externe Anschlüsse:
* I2C (4-Pin-Header)
* Analogeingang (3,5mm-Klinke)

Die Programmierung kann in allen Sprachen erfolgen, die für den ESP8266
verfügbar sind. Für Arduino-Programmierung in C / C++ steht diese Bibliothek zur
Verfügung, die die Programmierung stark vereinfacht.

## GPIO-Belegung

 0  Drehgeber Taster \
 1  TX \
 2  WS2812-Kette \
 3  RX \
 4  SDA \
 5  SCL \
12  Drehgeber Spur 0 \
13  Drehgeber Spur 1 \
14  DHT11 \
15  Lautsprecher \
16  LED / Reset

## Installation

Laden Sie unter [https://github.com/informatik-aalen/AALeC/releases](https://github.com/informatik-aalen/AALeC/releases)
die .ZIP Datei der aktuellen Version der Bibliothek herunter. Benutzen Sie den
Menüpunkt **.ZIP Bibliothek hinzufügen...** unter dem Menü **Sketch** -> 
**Bibliothek einbinden** in der Arduino IDE um die Bibliothek einzubinden.

## Abhängigkeiten

Diese Bibliothek hängt von den folgenden Bibliotheken ab:
* NeoPixelBus (Autor: *Makuna*, getestet mit Version 2.2.9)
* SimpleDHT (Autor: *Winlin*, getestet mit Version 1.06)
* ESP8266 and ESP32 Oled Driver for SSD1306 display (Autoren: *Daniel Eichhorn*, *Fabrice Weinberg*, geteste mit Version 3.2.7) 

Installieren Sie diese Bibliotheken bitte über den Bibliotheksverwalter in der Arduino IDE.

## Autoren

* **Winfried Bantel** - *Idee und Codierung*
* **Peter Kolb** - *Schaltplan, Layout, CAD, Mechanik*
* **Sebastian Stigler** - *Codierung*

## Lizenz

Das Projekt wird unter der MIT-Lizenz publiziert.

## Nachbau und Vertrieb

Nachbau ist explizit erwünscht. Fertigung und Vertrieb der Platine sind unter
Angabe der Hochschule Aalen erlaubt.


CHANGES.md
# Changelog

## Version 1.1.0
* Add rotate_changed. Fixes #7.

## Version 1.0.2
Sebastian Stigler (7):
* Bugfix for issue #1
* Merge pull request #2 from sebastianstigler/bugfix_invalid_led_nr
* Bugfix for issue #3
* Merge pull request #5 from sebastianstigler/bugfix_aalec_hangs_on_init
* Fixes #4
* Merge pull request #6 from sebastianstigler/enhancement_ISR_take_to_much_time

## Version 1.0.1
Sebastian Stigler (2):
* Add installation guide
* Bugfix: PIN_BUTTON set to correct value


## Version 1.0.0
Sebastian Stigler (6):
* The Library with an example for AALeC
* use define value
* Add more predifined colors and tones.\
  Add new set_rgb_strip variante.\
  Apply formating to code.
* Add Readme
* Add more exampes
* correct properties


src/AALeC.cpp
#include <Arduino.h>
#include <AALeC.h>
#include <NeoPixelBus.h>
#include <SimpleDHT.h>
#include "SSD1306.h"

c_AALeC aalec;


void AALeC_ISR_DREH() {
  static int16_t val, z, n;
  const static int8_t delta[4][4] = {
    {0, 1, 2, 0},
    {0, 1, 0, 3},
    {0, 0, 2, 3},
    {0, 1, 2, 3}
  },
  lambda[4][4] = {
    {  0,  1, -1,  0},
    { -1,  0,  0,  1},
    {  1,  0,  0, -1},
    {  0, -1,  1,  0}
  };
  noInterrupts();
  val = ~GPIO_REG_READ(GPIO_IN_ADDRESS) >> PIN_ENCODER_TRACK_1 & 3;
  n += lambda[z][val], z = delta[z][val];
  if (z == 0) {
    if (n == 4)
      aalec.drehgeber_int--;
    else if (n == -4)
      aalec.drehgeber_int++;
    n = 0;
  }
  interrupts();
}


c_AALeC::c_AALeC() {
}


void c_AALeC::init() {
  pinMode(PIN_BEEPER, OUTPUT);
  pinMode(PIN_LED_RESET, OUTPUT);
  pinMode(PIN_BUTTON, INPUT_PULLUP);
  pinMode(PIN_ENCODER_TRACK_1, INPUT_PULLUP);
  pinMode(PIN_ENCODER_TRACK_2, INPUT_PULLUP);

  GPIO_REG_WRITE(GPIO_OUT_W1TC_ADDRESS, 0xFFFF);
  attachInterrupt(PIN_ENCODER_TRACK_1, AALeC_ISR_DREH, CHANGE);
  attachInterrupt(PIN_ENCODER_TRACK_2, AALeC_ISR_DREH, CHANGE);

  Serial.begin(115200);
  while (!Serial);
  delay(100);
  Serial.println("AALeC init");
  Serial.flush();

  Wire.begin();
  strip = new NeoPixelBus<NeoRgbFeature, NeoEsp8266Uart800KbpsMethod>(3, PIN_RGB_STRIP);
  strip->Begin();
  strip->ClearTo(RgbColor(0, 0, 0));
  strip->Show();

  display.init();
  display.flipScreenVertically();
  display.setFont(ArialMT_Plain_10);
}


void c_AALeC::set_rgb_strip(int led, uint8_t r, uint8_t g, uint8_t b) {
  set_rgb_strip(led, RgbColor(r, g, b));
}


void c_AALeC::set_rgb_strip(int led, const RgbColor & c) {
    if (led >= 0 && led < 3) {
        strip->SetPixelColor(led, c);
        strip->Show();
    }
}


void c_AALeC::set_rgb_strip(int led, unsigned int x) {
  uint16_t r = x % 768, g = (x + 256) % 768, b = (x + 512) % 768;
  RgbColor c(
    (r < 256) ? r : ((r < 512) ? 511 - r : 0),
    (g < 256) ? g : ((g < 512) ? 511 - g : 0),
    (b < 256) ? r : ((b < 512) ? 511 - b : 0)
  );
  set_rgb_strip(led, c);
}


void c_AALeC::set_rgb_strip(const RgbColor & c0, const RgbColor & c1, const RgbColor & c2) {
  strip->SetPixelColor(0, c0);
  strip->SetPixelColor(1, c1);
  strip->SetPixelColor(2, c2);
  strip->Show();
}


void c_AALeC::set_led(int a) {
  digitalWrite(PIN_LED_RESET, a);
}


int c_AALeC::get_led() {
  return digitalRead(PIN_LED_RESET);
}


void c_AALeC::dim_led(int a) {
  analogWrite(PIN_LED_RESET, a);
}


int c_AALeC::get_button() {
  return  ~GPIO_REG_READ(GPIO_IN_ADDRESS)  & 1;
}


int c_AALeC::get_rotate() {
  return drehgeber_int;
}


int c_AALeC::rotate_changed() {
  int rc = 0;
  if (drehgeber_int !=drehgeber_int_alt)
    rc = 1, drehgeber_int_alt = drehgeber_int;
  return rc;
}


void c_AALeC::reset_rotate() {
  drehgeber_int = 0;
}


uint8_t c_AALeC::get_temp() {
  dht11_mess();
  return temp_int;
}


uint8_t c_AALeC::get_humidity() {
  dht11_mess();
  return hum_int;
}


uint16_t c_AALeC::get_analog() {
  return analogRead(A0);
}


void c_AALeC::clear_display() {
  display.clear();
}


void c_AALeC::play(unsigned int f) {
  if (f)
    tone(PIN_BEEPER, f);
  else
    noTone(PIN_BEEPER);
}


void c_AALeC::play(unsigned int f, unsigned int dur) {
  tone(PIN_BEEPER, f, dur);
}


void c_AALeC::dht11_mess() {
  uint32_t t = millis();
  if (t - last_dht11 >= 1000) {
    dht11.read(PIN_DHT11, &temp_int, &hum_int, NULL);
    last_dht11 = t;
  }
}


src/AALeC.h
#ifndef __AALEC_H
#define __AALEC_H
#include <NeoPixelBus.h>
#include <SimpleDHT.h>
#include "SSD1306.h"

#define PIN_BUTTON           0
#define PIN_TX               1
#define PIN_RGB_STRIP        2
#define PIN_RX               3
#define PIN_SDA              4
#define PIN_SDC              5
#define PIN_ENCODER_TRACK_1 12
#define PIN_ENCODER_TRACK_2 13
#define PIN_DHT11           14
#define PIN_BEEPER          15
#define PIN_LED_RESET       16

/* Predefined colors for the three rgb LED's */
const RgbColor c_red = RgbColor(200, 0, 0),
               c_green = RgbColor(0, 200, 0),
               c_blue = RgbColor(0, 0, 200),
               c_yellow = RgbColor(100, 100, 0),
               c_white = RgbColor(66, 66, 66),
               c_cyan = RgbColor(0, 100, 100),
               c_purple = RgbColor(100, 0, 100),
               c_off = RgbColor(0, 0, 0);


/* Some predefined frequencies of tones for the beeper */
enum {t_off = 0, t_c_1 = 262, t_d_1 = 294, t_e_1 = 330, t_f_1 = 349,
      t_g_1 = 392, t_a_1 = 440, t_h_1 = 494, t_c_2 = 523, t_a_2 = 880
     };

class c_AALeC {
  public:
    c_AALeC();
    void init();

    /* Methods for the three rgb LED's */
    void set_rgb_strip(int, uint8_t, uint8_t, uint8_t);
    void set_rgb_strip(int, const RgbColor &);
    void set_rgb_strip(int, unsigned int);
    void set_rgb_strip(const RgbColor &, const RgbColor &, const RgbColor &);

    /* Methods for the single LED */
    void set_led(int);
    int get_led();
    void dim_led(int);

    /* Method for the press button on the encoder */
    int get_button();

    /* Methods for the value of the encoder */
    int get_rotate();
    int rotate_changed();
    void reset_rotate();

    /* Methods for the temperature and humidity sensor */
    uint8_t get_temp();
    uint8_t get_humidity();

    /* Method for the potentiomether */
    uint16_t get_analog();

    /* Methods for the display */
    template <class T>    void print_line(unsigned int, const T &);
    void clear_display();

    /* Methods for the beeper */
    void play(unsigned int freq);
    void play(unsigned int freq, unsigned int dur);

    /* Interupt service routine for the encoder */
    friend void AALeC_ISR_DREH();

    /* Attribut for the display */
    SSD1306  display = SSD1306  (0x3c, 4, 5);

  private:
    void dht11_mess();

    int drehgeber_int = 0, drehgeber_int_alt = 0;
    NeoPixelBus<NeoRgbFeature, NeoEsp8266Uart800KbpsMethod> * strip;
    SimpleDHT11 dht11;
    uint8_t temp_int, hum_int;
    uint32_t last_dht11 = -1010;
} ;

template <class T> void c_AALeC::print_line(unsigned int line, const T & s) {
  display.setColor(BLACK);
  display.fillRect(0, ((line - 1) % 5) * 12, 128, 12);
  display.setColor(WHITE);
  display.drawString(0, ((line - 1) % 5) * 12 ,  String(s));
  display.display();
}

/* Default instance of the c_AALeC class */
extern c_AALeC  aalec;

#endif

